src/main/java/org/opensearch/ad/task/ADHCBatchTaskCache.java

```diff
@@ -220,7 +218,8 @@ public boolean hasRunningEntity() {
     }
 
     public boolean removeRunningEntity(String entity) {
-        return this.runningEntities.remove(entity);
+        // entity may still in temp queue, so need to remove from both temp and running queues
```

Comment:
Can you explain how the entity could be in either temp or running queue? Doesn't `moveToRunningEntity` remove such entity from the temp queue on line 170?

Reply:
In normal case, the entity will be moved to running queue if entity task dispatched to worker node successfully. If failed to dispatch to worker node, it will still stay in temp queue, check `ADBatchTaskRunner#workerNodeResponseListener`. Then will send entity task done message to coordinating node to move to pending queue if exception is retryable or remove entity from cache if not retryable.

Reply:
Could you add these comments?  They are really helpful.

Also, for HC detector, you will dispatch both detector level and task level tasks to worker nodes?  If yes, what's purpose of dispatching a detector level task to a worker node?  Could you point me any code link?

Reply:
For HC detector, will not dispatch detector level task to worker node, only dispatch entity level task. Check `ADBatchTaskRunner#forwardOrExecuteADTask`,  line 553 creates new entity task , `ADTask adEntityTask = new ADTask.Builder()`, line 578 forward entity task to worker node, `forwardOrExecuteEntityTask(adEntityTask, transportService, workerNodeResponseListener);`